clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -1029,15 +1062,37 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                      shard->file_type_)) {
             if (type_is_read(memref.data.type)) {
                 // Skip D-filtered traces which don't have every load or store records.
-                report_if_false(shard, shard->expected_read_records_ > 0,
-                                "Too many read records");
+                report_if_false(
+                    shard,
+                    shard->expected_read_records_ >
+                        0 IF_X86(
+                            // iret pops bunch of data from the stack at interrupt
+                            // returns. Read count cannot be determined using the
+                            // decoder. System call trace templates collected on
+                            // QEMU would show all reads.
+                            || (shard->between_kernel_syscall_trace_markers_ &&
+                                shard->prev_instr_.decoding.opcode == OP_iret)
+                            // Xrstor does multiple reads. Read count cannot be
+                            // determined using the decoder. System call trace
+                            // templates collected on QEMU would show all reads.
+                            || (shard->between_kernel_syscall_trace_markers_ &&
+                                shard->prev_instr_.decoding.is_xrstor)),
```

Comment:
If we have the opcode stored do we need these separate is_ bools?  Seems like you'd have one or the other, not both: redundant.

Reply:
For some we have routines like instr_is_xrstor, instr_is_xsave. I wanted to re-use those where possible.

Reply:
OK.